src/main/java/io/cryostat/recordings/RecordingMetadataManager.java

```diff
@@ -43,113 +43,388 @@
+                                        try {
+                                            if (isArchivedRecording(recordingName)) {
+                                                RecordingArchiveHelper archiveHelper =
+                                                        archiveHelperProvider.get();
+                                                Path recordingPath =
+                                                        archiveHelper
+                                                                .getRecordingPath(recordingName)
+                                                                .get();
+                                                String subdirectoryName =
+                                                        recordingPath
+                                                                .getParent()
+                                                                .getFileName()
+                                                                .toString();
+                                                String newTargetId =
+                                                        new String(
+                                                                base32.decode(subdirectoryName),
+                                                                StandardCharsets.UTF_8);
+                                                logger.info(
+                                                        "Found metadata corresponding to archived"
+                                                                + " recording: {}",
+                                                        recordingName);
+                                                setRecordingMetadata(
+                                                        new ConnectionDescriptor(newTargetId),
+                                                        recordingName,
+                                                        new Metadata(srm.getLabels()));
+                                            } else {
+                                                logger.warn(
+                                                        "Found metadata for lost archived"
+                                                                + " recording: {}",
+                                                        recordingName);
+                                            }
+                                        } catch (InterruptedException | ExecutionException e) {
+                                            logger.error(e);
+                                        }
+
+                                    } else {
+                                        try {
+                                            ConnectionDescriptor cd =
+                                                    new ConnectionDescriptor(targetId);
+                                            if (targetRecordingExists(cd, recordingName).get()) {
```

Comment:
Checking if an active recording exists can be a pretty slow operation. Can this check and staleness pruning be done in parallel? I think it should also be done async from startup, ie we should allow the verticle to be considered deployed even while this work is ongoing. If there were a lot of active recordings during the last lifeycle, ex. there was an Automated Rule activating recordings on a large number of targets, this could take really substantial time to complete.

Reply:
Hm... I'm not sure what you mean by "async from startup". Isn't the `RecordingMetadataManager` already async as a worker verticle? Or did you want to move the check and stateless pruning logic separately to some other module where these other non-verticles are started? e.g.
```
    @Override
    public void start(Promise<Void> future) {
        logger.trace("env: {}", environment.getEnv().toString());

        try {
            client.credentialsManager().migrate();
            client.credentialsManager().load();
            client.ruleRegistry().loadRules();
            # client.someNewModule().pruneOldMetadata();
        } catch (Exception e) {
            logger.error(e);
            future.fail(e);
            return;
        }
```

Reply:
I just mean that the verticle starts up with this:

```java
 public void start(Promise<Void> future)
```

This pruning step is deeply nested within some logic that is all running synchronously, and at the very end the `future.complete();` is what finishes the verticle's startup. I think that also means subsequent verticles would not be deployed until this completes now that the main method in `Cryostat.java` uses a serial startup sequence.

Reply:
Right, I forgot that the verticle startup is synchronous now. I tried just completing the future earlier by calling `future.complete()` before the pruning logic, but I don't think that works. I look into it tomorrow.

Reply:
I think you can just wrap the pruning logic inside a `vertx.executeBlocking()` or an `executor.submit()` to offload it onto another thread and allow the one performing the deployment to continue on.